Finish implementation of optional attributes

siku2 opened the pull request:
#### Description

See the linked pull request and issue for context.

Continuation of #1433
Closes #903

#### Details

- When no optional attributes are involved, more optimised code is generated.
- Uses `Attributes::IndexMap` instead of `Attributes::Vec` when optional attributes are present.
- Cleaned up a bunch of macro code.

#### Checklist:

- [x] I have run `./ci/run_stable_checks.sh`
- [x] I have reviewed my own code
- [x] I have added tests

teymour-aldridge commented on docs/concepts/html/elements.md:
````diff
@@ -89,6 +89,22 @@ html! {
+}
+```
+
+If the attribute is set to `None`, it will behave as though it wasn't set.
````
```suggestion
If the attribute is set to `None`, it attribute won't be set in the DOM.
```

teymour-aldridge replied:
It's *very* minor.

siku2 replied:
That line stood out to me as well, thanks.

siku2 commented:
Huh, what's going on with the checks...

teymour-aldridge commented:
Maybe with hit peak concurrency?
(Github Actions limits the number of concurrently running actions)

> On 31 Aug 2020, at 15:50, Simon <notifications@github.com> wrote:
> 
> 
> Huh, what's going on with the checks...
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/yewstack/yew/pull/1537#issuecomment-683827743>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKFSTPLP2AVF77LYQD2LZUTSDO2BZANCNFSM4QQS4AFA>.
>

siku2 commented:
@jstarry, do you have some time to review this? I had to make quite a few changes to the original PR.

jstarry commented:
Sure, I'll take a look tomorrow!

jstarry reviewed (commented):
Looks really solid! Love the cleanup, and thanks for fixing up the vtag API and that class macro bug.

Just one question, why default to using `IndexMap` when using optional attributes?

alexschrod commented on yew-macro/tests/macro/html-tag-fail.stderr:
```diff
@@ -265,44 +257,43 @@ error[E0277]: the trait bound `std::borrow::Cow<'static, str>: std::convert::Fro
-   = note: required by `std::convert::From::from`
-   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
+   = note: required because of the requirements on the impl of `std::convert::Into<std::borrow::Cow<'static, str>>` for `{integer}`
+   = note: required by `std::convert::Into::into`
```
What happened here? Maybe it's a nitpick on my part, but I feel like this error was a lot less messy when requiring `From` rather than `Into`.

siku2 replied:
This isn't something I touched in this PR, this change was part of #1447 and is already in:
https://github.com/yewstack/yew/blob/master/yew-macro/tests/macro/html-tag-fail.stderr#L236-L237

Gotta love diff quirks like this.

Anyhow, it does make more sense to accept `Into` than `From` because of the additional flexibility. The slightly more verbose error message is an acceptable trade-off in my opinion.

alexschrod replied:
Ah yeah, you're right, there can be more `Into` implementations than `From` implementations, since `From` implies `Into`, but not vice versa. Unfortunate about the error message, but a worthwhile trade-off, I agree.

siku2 commented:
> Just one question, why default to using `IndexMap` when using optional attributes?

@jstarry 
If an optional attribute switches from `None` to `Some` or vice versa the length of the attributes changes. When diffing attribute vectors that don't share the same length it calls [`diff_incompatible`](https://github.com/yewstack/yew/blob/18b9086bc9ca489a311dc07907ed2ee70527e751/yew/src/virtual_dom/vtag.rs#L299) to compare them which converts both vectors into `HashMap`s first. This isn't needed when using  `IndexMap`.
I haven't benchmarked this so I can't say for sure that this is more performant but I'm fairly certain that it is.

jstarry commented:
> If an optional attribute switches from None to Some or vice versa the length of the attributes changes. When diffing attribute vectors that don't share the same length, diff_incompatible is called which converts both vectors into HashMaps first. This isn't needed when using IndexMap.

Good point. What if we option wrapped all of the attribute values to maintain order? 

`Vec<(&'static str, Option<Cow<'static, str>>)>`

siku2 commented:
That would work, yeah. Should even make it possible to clean up the code significantly.
The only downside I can think of is that this would slightly slow down the non-optional attribute path because of the runtime overhead of `Option`.

siku2 commented:
@jstarry Well that was **A LOT** more painful than I thought but it's done and includes some much needed clean up too.

siku2 commented:
Since I changed the attribute diffing and the benchmarks, here's what it looks like now:
```js
changed last:
 vec-vec:          0s (R²=0.993, 1233946 iterations in 122 samples)
 map-map:          0s (R²=0.999, 1357342 iterations in 123 samples)
 vec-map:         3ns (R²=0.780, 357422 iterations in 109 samples)
 map-vec:         2ns (R²=0.889, 295388 iterations in 107 samples)
changed middle:
 vec-vec:         1ns (R²=0.898, 1019788 iterations in 120 samples)
 map-map:          0s (R²=0.997, 1357342 iterations in 123 samples)
 vec-map:         2ns (R²=0.984, 575638 iterations in 114 samples)
 map-vec:         2ns (R²=0.999, 432483 iterations in 111 samples)
changed first:
 vec-vec:         1ns (R²=0.999, 1019788 iterations in 120 samples)
 map-map:         1ns (R²=0.998, 1019788 iterations in 120 samples)
 vec-map:         3ns (R²=0.912, 393165 iterations in 110 samples)
 map-vec:         5ns (R²=0.990, 183408 iterations in 102 samples)
reordered:
 vec-vec:         6ns (R²=0.967, 137793 iterations in 99 samples)
 map-map:          0s (R²=0.965, 4259946 iterations in 135 samples)
 vec-map:         2ns (R²=0.969, 432483 iterations in 111 samples)
 map-vec:         3ns (R²=0.971, 324928 iterations in 108 samples)
added to start:
 vec-vec:        10ns (R²=0.981, 94110 iterations in 95 samples)
 map-map:          0s (R²=0.976, 2909596 iterations in 131 samples)
 vec-map:         3ns (R²=0.987, 295388 iterations in 107 samples)
 map-vec:         2ns (R²=0.986, 432483 iterations in 111 samples)
removed from start:
 vec-vec:        10ns (R²=0.977, 113876 iterations in 97 samples)
 map-map:          0s (R²=0.991, 2909596 iterations in 131 samples)
 vec-map:         2ns (R²=0.984, 432483 iterations in 111 samples)
 map-vec:         3ns (R²=0.991, 295388 iterations in 107 samples)
added to end:
 vec-vec:         1ns (R²=0.980, 696525 iterations in 116 samples)
 map-map:         1ns (R²=0.997, 927079 iterations in 119 samples)
 vec-map:         6ns (R²=0.926, 201750 iterations in 103 samples)
 map-vec:         7ns (R²=0.991, 137793 iterations in 99 samples)
removed from end:
 vec-vec:         2ns (R²=0.993, 432483 iterations in 111 samples)
 map-map:         2ns (R²=0.996, 432483 iterations in 111 samples)
 vec-map:         4ns (R²=0.929, 201750 iterations in 103 samples)
 map-vec:         3ns (R²=0.972, 244120 iterations in 105 samples)
equal:
 vec-vec:          0s (R²=0.991, 2645086 iterations in 130 samples)
 map-map:          0s (R²=0.977, 2909596 iterations in 131 samples)
 vec-map:         1ns (R²=0.997, 575638 iterations in 114 samples)
 map-vec:         1ns (R²=0.962, 575638 iterations in 114 samples)
```

siku2 commented:
In hindsight it probably would've been better to do the switch to `Optional<Cow<'static, str>>` in a separate PR but I'm pretty happy with how this turned out otherwise.
Would you like to take another look, @jstarry?

thedodd commented:
Overall the code is looking really great. I haven't reviewed the Yew code base much, but I was happy to see `trybuild` being used, it's a lifesaver for sure. 

I have a repo using the code over in #1433 (@alexschrod's original PR code). I'll cut the repo over to this branch and will report back if I see any issues.

siku2 commented:
I was a bit bothered by the outliers in the vec-vec case so I decided to optimize it a bit.
These slow cases happen when the two attribute vectors don't have the keys in the same order. This is true for the "reordered", "added to start", and "removed from start" benchmarks.
Here are the results:

```js
changed last:
 vec-vec:         1ns (R²=0.977, 1357342 iterations in 123 samples)
 map-map:         1ns (R²=0.993, 1642386 iterations in 125 samples)
 vec-map:         4ns (R²=0.876, 523306 iterations in 113 samples)
 map-vec:         2ns (R²=0.999, 842798 iterations in 118 samples)
changed middle:
 vec-vec:         1ns (R²=1.000, 1987290 iterations in 127 samples)
 map-map:         1ns (R²=1.000, 2186020 iterations in 128 samples)
 vec-map:         2ns (R²=1.000, 927079 iterations in 119 samples)
 map-vec:         2ns (R²=1.000, 927079 iterations in 119 samples)
changed first:
 vec-vec:          0s (R²=1.000, 2404623 iterations in 129 samples)
 map-map:          0s (R²=1.000, 2404623 iterations in 129 samples)
 vec-map:         2ns (R²=1.000, 1019788 iterations in 120 samples)
 map-vec:         2ns (R²=1.000, 1019788 iterations in 120 samples)
reordered:
 vec-vec:         2ns (R²=1.000, 927079 iterations in 119 samples)
 map-map:          0s (R²=0.999, 13369557 iterations in 147 samples)
 vec-map:         1ns (R²=0.998, 1493077 iterations in 124 samples)
 map-vec:         1ns (R²=1.000, 1493077 iterations in 124 samples)
added to start:
 vec-vec:         3ns (R²=1.000, 575638 iterations in 114 samples)
 map-map:          0s (R²=1.000, 12154142 iterations in 146 samples)
 vec-map:         1ns (R²=1.000, 1357342 iterations in 123 samples)
 map-vec:         1ns (R²=1.000, 1357342 iterations in 123 samples)
removed from start:
 vec-vec:         3ns (R²=0.999, 696525 iterations in 116 samples)
 map-map:          0s (R²=1.000, 13369557 iterations in 147 samples)
 vec-map:         1ns (R²=0.980, 1493077 iterations in 124 samples)
 map-vec:         2ns (R²=0.971, 927079 iterations in 119 samples)
added to end:
 vec-vec:         1ns (R²=0.997, 1987290 iterations in 127 samples)
 map-map:         1ns (R²=0.997, 1806626 iterations in 126 samples)
 vec-map:         2ns (R²=0.998, 842798 iterations in 118 samples)
 map-vec:         2ns (R²=0.999, 927079 iterations in 119 samples)
removed from end:
 vec-vec:          0s (R²=1.000, 2909596 iterations in 131 samples)
 map-map:          0s (R²=1.000, 2909596 iterations in 131 samples)
 vec-map:         2ns (R²=0.998, 1019788 iterations in 120 samples)
 map-vec:         2ns (R²=0.997, 927079 iterations in 119 samples)
equal:
 vec-vec:          0s (R²=0.999, 5154537 iterations in 137 samples)
 map-map:          0s (R²=0.943, 5669992 iterations in 138 samples)
 vec-map:         1ns (R²=0.999, 1233946 iterations in 122 samples)
 map-vec:         1ns (R²=1.000, 1233946 iterations in 122 samples)
```

jstarry reviewed (commented):
Agreed that handling duplicate vec attributes can be skipped! Can you add a comment on the `__macro_push_attribute` function that non-unique attributes will cause erratic behaviour?

siku2 commented on yew/src/virtual_dom/mod.rs:
```diff
@@ -60,70 +60,249 @@ impl fmt::Debug for dyn Listener {
+                    }
+
+                    // finally, we're left with the attributes that are actually new.
+                    out.extend(added.into_iter().map(|(k, v)| Patch::Add(k, v)));
```
Does it need to be deterministic? It shouldn't matter in which order the patches are applied.
(If it matters then we can't use this algorithm at all because additions will always be performed after `Replace` and `Remove`)

jstarry replied:
Ah, you're right. Order doesn't matter. I was confused 😅

siku2 commented on yew-macro/src/html_tree/html_tag/mod.rs:
```diff
@@ -137,95 +133,192 @@ impl ToTokens for HtmlTag {
+                            let ident = Ident::new("__yew_listener", name.span());
+                            let listener = generate_listener_code(name, &ident);
+                            quote_spanned! {value.span()=>
+                                let #ident = ::std::option::Option::map(#value, |#ident| {
+                                    #listener
+                                });
+                                if let ::std::option::Option::Some(#ident) = #ident {
+                                    #vtag.add_listener(#ident);
+                                };
```
Yes, but it generates a worse error message. Instead of saying "expected an Option, found something else" it will say "expected <type of value>, found <fully qualified Option type>"

jstarry replied:
Ah got it, sounds good!

siku2 commented on yew/src/virtual_dom/mod.rs:
```diff
@@ -60,70 +60,249 @@ impl fmt::Debug for dyn Listener {
+                )
+            }
+            (Self::IndexMap(new), Self::IndexMap(old)) => {
+                let new_iter = new.iter().map(|(k, v)| (*k, v.as_ref()));
```
Yeah, it's about lifetimes. Normally the lifetime for the value `&'a str` comes from `&'a IndexMap<_, _>` but in the vec-map case we're converting the vector into an owned `IndexMap` over references to the vector's data so the lifetime comes from `IndexMap<_, &'a str>`.

The alternative was to create a trait or to duplicate the code. Both of which would've been more verbose

jstarry replied:
Got it, this approach works for me

jstarry replied:
Can you please add a comment on the `new_iter` arg explaining this?

siku2 commented on yew-macro/src/html_tree/html_tag/mod.rs:
```diff
@@ -252,34 +345,23 @@ impl ToTokens for HtmlTag {
-                    #vtag.add_children(#children);
-                };
+                #set_attributes
+                #set_booleans
+                #set_classes
```
great idea to distinguish those, yes!

siku2 commented on yew-macro/src/html_tree/html_tag/mod.rs:
```diff
@@ -1,14 +1,13 @@
-use proc_macro2::{Delimiter, Span};
+use proc_macro2::{Delimiter, Span, TokenStream};
 use quote::{quote, quote_spanned, ToTokens};
+use stringify::Stringify;
```
I had the same impression too

siku2 commented on yew-macro/src/stringify.rs:
```diff
@@ -1,68 +1,91 @@
-        Expr::Lit(l) => try_stringify_lit(&l.lit),
-        _ => None,
+/// Stringify a dynamic value at runtime.
+pub fn stringify_dynamic(src: impl ToTokens) -> TokenStream {
```
Hmm... I agree but I have to wonder what made me change it in the first place.

siku2 replied:
Especially because `stringify_optiona_at_runtime` still uses the old pattern. This was probably a change I forgot to undo while trying to improve the stringify API.

siku2 commented on yew-macro/src/html_tree/html_tag/mod.rs:
```diff
@@ -137,95 +133,192 @@ impl ToTokens for HtmlTag {
+                    quote_spanned! {value.span()=> {
+                        if #value {
+                            #vtag.__macro_push_attribute(::yew::virtual_dom::PositionalAttr::new(#label_str, #sr));
+                        };
```
Ah, I totally forgot about the entire reason I went through the trouble of changing it :laughing:.
Thanks for reminding me!

siku2 commented on yew-macro/src/stringify.rs:
```diff
@@ -1,68 +1,91 @@
-    }};
+/// Create `Cow<'static, str>` construction calls.
+///
+/// This is deliberately not implemented for strings to preserve spans.
```
When you use a string literal in `quote!` it doesn't have a `Span` attached to it so if there's an error involving it the span will point to the entire call site. The `Stringify` trait isn't implemented for any `String` type to avoid that kind of mistake

jstarry replied:
I see, thanks for the explanation!

siku2 commented on yew-macro/src/html_tree/html_tag/mod.rs:
```diff
@@ -137,95 +133,192 @@ impl ToTokens for HtmlTag {
+                    } else {
+                        let sr = lit.stringify();
+                        Some(quote! {
+                            #vtag.__macro_push_attribute(::yew::virtual_dom::PositionalAttr::new("class", #sr));
```
That's already checked by `lit.value().is_empty()`.

jstarry replied:
Ah cool, missed that. Thanks!

siku2 commented on yew-macro/src/html_tree/html_tag/mod.rs:
```diff
@@ -137,95 +133,192 @@ impl ToTokens for HtmlTag {
+                     question_mark,
+                     value,
+                 }| {
+                    generate_optional_attribute_code(
```
this is the only remaining usage of the function so I just inlined it

siku2 commented:
I think this should be good to go now.

jstarry commented on yew-macro/src/html_tree/html_tag/mod.rs:
```diff
@@ -137,95 +133,205 @@ impl ToTokens for HtmlTag {
+            Some(ClassesForm::Single(classes)) => match classes.try_into_lit() {
+                Some(lit) => {
+                    if lit.value().is_empty() {
+                        None
```
Should we push a placeholder attribute here?

siku2 replied:
The reason I didn't is because this is the literal case (i.e. `html! { <div class="weird-button"></div> }`) so the value never changes. The only reason for the `is_empty()` check is so that `html! { <div class=""></div> }` doesn't add the class attribute (though I have to wonder whether we shouldn't have that be a warning because I don't see this ever being useful).

jstarry replied:
Ah yeah great point

jstarry reviewed (approved):
Awesome! Just one question :)

Mingun commented on yew-macro/tests/macro/html-tag-fail.stderr:
```diff
@@ -136,6 +136,30 @@ error: this dynamic tag is missing an expression block defining its value
+error: the `class` attribute does not support being used as an optional attribute
+  --> $DIR/html-tag-fail.rs:59:20
+   |
+59 |     html! { <input class?=() /> };
```
Why `class` can't be used as optional attribute?

siku2 replied:
`class` already supports `Option<String>` as a value.

Mingun commented on yew-macro/src/html_tree/html_component.rs:
```diff
@@ -414,6 +414,13 @@ impl Parse for Props {
+                    "optional attributes are only supported on HTML tags. Yew components can use `Option<T>` properties to accomplish the same thing.",
+                ));
+            }
+
```
That reminds me, why I'm not in flavor of `?` syntax. Thats create unnecessary difference between built-in and custom components with no good reason

siku2 replied:
That difference is in no way unnecessary. All attributes are technically optional for normal html elements but the same cannot be said for components where the component type controls which props are optional and which are required.

You could argue that the syntax should work for optional Component props but that's a different issue.

jstarry replied:
@siku2 could we avoid using `?` if we did `Into<Option<...>>`? Or does it run into issues?

siku2 replied:
> @siku2 could we avoid using `?` if we did `Into<Option<...>>`? Or does it run into issues?

@jstarry
I talked about that here: https://github.com/yewstack/yew/issues/903#issuecomment-680808862.
The second point no longer stands because we unified the runtime interface.

Do you feel it would be more ergonomic to drop the explicit syntax in favor of `Into<Option<_>>`?

Mingun replied:
> Alternatively, we can also adjust the behaviour of fields marked with `prop_or*` so that they take `Into<Option<T>>` values where `None` acts the same way as if the prop wasn't given.

Yes, that is exactly what I prefer to see

siku2 replied:
I opened a new issue for this: #1550
